uxarray/grid.py

```diff
@@ -372,6 +373,82 @@ def integrate(self, var_ds, quadrature_rule="triangular", order=4):
+        face.
+
+        Notes
+        -----
+        This function will add `Grid.ds.Mesh2_face_edges` to the `Grid` class, which is an integer
+        DataArray of size (nMesh2_face, MaxNumNodesPerFace)
+        """
+        mesh2_face_nodes = self.Mesh2_face_nodes.values
+        n = self.nMesh2_face
+        m = self.nMaxMesh2_face_nodes
+
+        # First identify the _FillValue in the mesh2_face_nodes
+        # We will unify the _FillValue used in this function as -1:
+        if np.isnan(_FillValue):
+            mesh2_face_nodes = np.nan_to_num(mesh2_face_nodes, nan=-1)
+        elif _FillValue is not -1:
+            mesh2_face_nodes[mesh2_face_nodes == _FillValue] = -1
+
+        # Then do the padding for each face to close the polygon
+        closed = np.full((n, m + 1), -1, dtype=np.intp)
+        closed[:, :-1] = np.array(mesh2_face_nodes, dtype=np.intp)
+        # We only want the index of first occurrence of -1
+        first_fill_value_index = np.argmax(closed == -1, axis=1)
+        first_node = mesh2_face_nodes[:, 0]
+
+        # Now replace the first -1 at each row to be the first node
+        pad_results = np.array(
+            list(
+                map(self.__replace_fill_val, closed, first_fill_value_index,
+                    first_node)))
+
+        # Now create the edge_node_connectivity
+        mesh2_edge_nodes = np.empty((n * m, 2), dtype=np.intp)
+        mesh2_edge_nodes[:, 0] = pad_results[:, :-1].ravel()
+        mesh2_edge_nodes[:, 1] = pad_results[:, 1:].ravel()
+
+        # Clean up the invalid edge (same node to same node except the [-1, -1] edge)
+        valid_mask = np.array(
+            list(
+                map(
+                    lambda edge: not (edge[0] == edge[1] and not np.array_equal(
+                        edge, np.array([-1, -1]))), mesh2_edge_nodes)))
+        mesh2_edge_nodes = mesh2_edge_nodes[valid_mask]
+
+        # Find the unique edge
+        mesh2_edge_nodes.sort(axis=1)
+        mesh2_edge_node_copy, inverse_indices = np.unique(ar=mesh2_edge_nodes,
+                                                          return_inverse=True,
+                                                          axis=0)
+
+        # In mesh2_edge_nodes, we want to remove all dummy edges (edge that has "-1" node index)
+        # But we want to preserve that in our mesh2_face_edges so make the datarray has the same dimensions
+        cumprod = (mesh2_edge_node_copy[:, 0] + 1) * \
+                  (mesh2_edge_node_copy[:, 1] + 1) # In this case, former 0 will be positive while former -1 will be 0
+        has_fill_value = np.ma.masked_where(cumprod != 0, cumprod).mask
+        mesh2_edge_nodes = mesh2_edge_node_copy[has_fill_value] if isinstance(
+            has_fill_value, np.ndarray) else mesh2_edge_node_copy
+        inverse_indices = inverse_indices.reshape(n, m)
+        mesh2_face_edges = mesh2_edge_node_copy[inverse_indices]
```

Comment:
Would you mind giving me an overview of what is happening in this section?

Reply:
The basic idea behind this is our `mesh2_face_nodes` will have Filled_values

> uxarray.Grid.Mesh2_face_nodes: integer DataArray of size (nMesh2_face, MaxNumNodesPerFace)
> A DataArray of indices for each face node, corresponding to coordinates in uxarray.Grid.node_*. Faces can have arbitrary length, with _FillValue=-1 used when faces have fewer nodes than MaxNumNodesPerFace. Nodes are in counter-clockwise order.

And also the fact that our `mesh2_face_edges` has fixed dimensions

> uxarray.Grid.Mesh2_face_edges: integer DataArray of size (nMesh2_face, MaxNumNodesPerFace) (optional)
> A DataArray of indices indicating edges that are neighboring each face.

So in order to incorporate Huite's implementation from XUGRID, we need to do the following
1. For the `mesh2_edge_nodes`, we don't want it to contain any "dummy edges" (the edge that has `-1` as its node)
2. However, for the  `Grid.Mesh2_face_edges`, we need to preserve the dummy edges to make sure it has the same dimensions, For example,
A face with node [0 ,3 5 , -1] means it's a triangle [(0, 3), (3, 5), (5, 0)], however, this will make it only have a size of 3, so we need it to look like [(0, 3), (3, 5), (5, 0), (0, -1)]. For the last `dummy edge`, we don't care what's the valid node number, as long as this edge has `-1` as its node, it will be considered as the  `dummy edge`

Therefore, after line 429, we have the unique edges set `mesh2_edge_node_copy` and it may look like this:

> [[0, 1],
> [1, 2],
> ...
> [1, -1]
> ...
> [0, -1]
> ]

And from line 435 to line line 438 we need to filter out all "valid edge" by removing the "invalid edge". AKA, we want to know the index of all edge that contains -1. Naturally, we will try `mesh2_edge_node_copy[0] * mesh2_edge_node_copy[1]` and get whatever index that's negative. But `0 * -1` will be equal to `0 * 1`(or any other thing). So we're going to increase every index by 1, which means 0 is now 1 and -1 will be 0. 

So we just need to find out
`mesh2_edge_node_copy[:, 0] * mesh2_edge_node_copy[:,1] != 0`, and that will be the valid edge index

Reply:
Another thing I am thinking about right now is how to make every face edge in the counter-clockwise order, @philipc2 I wonder if you have any idea about it? thanks

Reply:
Thank you for the explanation!  

You mentioned that

>And also the fact that our mesh2_face_edges has fixed dimensions

 > > uxarray.Grid.Mesh2_face_edges: integer DataArray of size (nMesh2_face, MaxNumNodesPerFace) (optional)
A DataArray of indices indicating edges that are neighboring each face.

In your implementation, `Mesh2_face_edges` has dimensions `[nMesh2_face, nMaxMesh2_face_nodes, 2]`. Can you explain why you have the extra dimension in `Mesh2_face_edges`?

Reply:
> Thank you for the explanation!
> 
> You mentioned that
> 
> > And also the fact that our mesh2_face_edges has fixed dimensions
> 
> > > uxarray.Grid.Mesh2_face_edges: integer DataArray of size (nMesh2_face, MaxNumNodesPerFace) (optional)
> > > A DataArray of indices indicating edges that are neighboring each face.
> 
> In your implementation, `Mesh2_face_edges` has dimensions `[nMesh2_face, nMaxMesh2_face_nodes, 2]`. Can you explain why you have the extra dimension in `Mesh2_face_edges`?

I guess it's because I am still not very familiar with the Xarray "dimension" attributes. I thought it refers to the size of each matrix dimensions

So the Mesh2_face_edges look like this

> [[0,1],[1,2],[2,3],[3,0],
>  [0,2],[2,4],[4,5],[5,0],
> [0,2],[2,4],[4,0],[-1,-1],
> ....
> ]

So it will have the `nMesh2_face` of rows, in each row, it will have the `nMaxMesh2_face_nodes` number of edges, and then inside each edge, the size of the edge array is 2.

Please let me know if you have any better idea about the data structure. Thanks!

Reply:
> The basic idea behind this is our `mesh2_face_nodes` will have Filled_values
> 
> > uxarray.Grid.Mesh2_face_nodes: integer DataArray of size (nMesh2_face, MaxNumNodesPerFace)
> > A DataArray of indices for each face node, corresponding to coordinates in uxarray.Grid.node_*. Faces can have arbitrary length, with _FillValue=-1 used when faces have fewer nodes than MaxNumNodesPerFace. Nodes are in counter-clockwise order.
> 
> And also the fact that our `mesh2_face_edges` has fixed dimensions
> 
> > uxarray.Grid.Mesh2_face_edges: integer DataArray of size (nMesh2_face, MaxNumNodesPerFace) (optional)
> > A DataArray of indices indicating edges that are neighboring each face.
> 
> So in order to incorporate Huite's implementation from XUGRID, we need to do the following
> 
> 1. For the `mesh2_edge_nodes`, we don't want it to contain any "dummy edges" (the edge that has `-1` as its node)
> 2. However, for the  `Grid.Mesh2_face_edges`, we need to preserve the dummy edges to make sure it has the same dimensions, For example,
>    A face with node [0 ,3 5 , -1] means it's a triangle [(0, 3), (3, 5), (5, 0)], however, this will make it only have a size of 3, so we need it to look like [(0, 3), (3, 5), (5, 0), (0, -1)]. For the last `dummy edge`, we don't care what's the valid node number, as long as this edge has `-1` as its node, it will be considered as the  `dummy edge`
> 
> Therefore, after line 429, we have the unique edges set `mesh2_edge_node_copy` and it may look like this:
> 
> > [[0, 1],
> > [1, 2],
> > ...
> > [1, -1]
> > ...
> > [0, -1]
> > ]
> 
> And from line 435 to line line 438 we need to filter out all "valid edge" by removing the "invalid edge". AKA, we want to know the index of all edge that contains -1. Naturally, we will try `mesh2_edge_node_copy[0] * mesh2_edge_node_copy[1]` and get whatever index that's negative. But `0 * -1` will be equal to `0 * 1`(or any other thing). So we're going to increase every index by 1, which means 0 is now 1 and -1 will be 0.
> 
> So we just need to find out `mesh2_edge_node_copy[:, 0] * mesh2_edge_node_copy[:,1] != 0`, and that will be the valid edge index

@philipc2 This is the description for why the FillValue = -1 will be important in my implementation here. Please let me know if you any question! Thanks!

Reply:
@hongyuchen1030 

Thank you! I'll give this another read through this week 😀